### PR TITLE
GPII-2732

### PR DESCRIPTION
**Pull request by danailbd**
This PR introduces the survey pop-up functionality which includes:
* A simple mocked local survey server (for dev purposes only)
* Communication between the survey server and the PSP main process over WebSocket
    * exchange of survey trigger definitions and survey pop-up config
    * new payload schema definition
    * third party dependency usage - [node-machine-id](https://github.com/automation-stack/node-machine-id)
* Implementation of a survey pop-up
    * electron BrowserWindow
    * embedding of Qualtrics survey inside a webview
* Simple time-based Rules Engine (triggers mechanism)
    * wrapping of a third party dependency - [json-rules-engine](https://github.com/cachecontrol/json-rules-engine) and directly using its provided conditions schema
    * frequent supply of app facts (e.g. keyedInBefore - time since keyed in) by the `factsManger`
    * time-based conditions checks

For further details please refer to the related tickets' comments and components documentation.

Related tickets:
* [GPII-2755](https://issues.gpii.net/browse/GPII-2755)
* [GPII-2758](https://issues.gpii.net/browse/GPII-2758)
* [GPII-2759](https://issues.gpii.net/browse/GPII-2759)
* [GPII-2760](https://issues.gpii.net/browse/GPII-2760)
* [GPII-2762](https://issues.gpii.net/browse/GPII-2762)
* [GPII-2763](https://issues.gpii.net/browse/GPII-2763)
* [GPII-2764](https://issues.gpii.net/browse/GPII-2764)
* [GPII-2776](https://issues.gpii.net/browse/GPII-2776): Triggers Mechanism - detailed description in comments

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -11,22 +11,28 @@
     "electron-packager": "8.5.1",
     "eslint-config-fluid": "1.0.0",
     "fluid-grunt-eslint": "18.1.2",
+    "gpii-survey-server": "git+https://github.com/danailbd/gpii-survey-server.git",
     "grunt": "1.0.1",
     "grunt-jsonlint": "1.1.0",
     "grunt-shell": "1.3.0",
+    "npm-run-all": "^4.1.2",
```
Comment (kaspermarkus) [inline package.json]: Should be specific version

File: src/main/app.js [hunk 2]
```diff
@@ -208,41 +284,37 @@ fluid.defaults("gpii.app", {
...
+                    onCreate: "{app}.events.onPSPReady.fire"
                 }
             }
-        },
-        networkCheck: { // Network check component to meet GPII-2349
-            type: "gpii.app.networkCheck"
         }
     },
     events: {
         onPrerequisitesReady: {
```
Comment (kaspermarkus) [inline src/main/app.js]: not sure if this makes sense, but perhaps renaming to `onPSPPrerequisitesReady` would make it clear what pre-requisites refers to

File: src/main/dialogManager.js [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+ * throughout the PSP application. All dialogs that can be manipulated by this
+ * component need to be registered as its subcomponents. If a dialog cannot be
+ * created when the `dialogManager` is created, it can be wrapped in a component
+ * which will take care of the dialogs creation when it is needed (for reference,
+ * see the `gpii.app.survey` component which wraps the `gpii.app.surveyDialog` and
+ * creates it via an event).
+ * In order to show/hide/close a dialog, the IoCSS selector of the particular
+ * component needs to be provided as an argument of the corresponding function.
+ */
+// XXX: Currently only the survey dialog is managed by the dialogManager. As a
```
Comment (kaspermarkus) [inline src/main/dialogManager.js:33]: Could you create a JIRA for this work (if we dont already have one)

File: src/main/dialogManager.js [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    var dialogs = fluid.queryIoCSelector(dialogManager, selector);
+    if (dialogs.length > 0) {
+        return dialogs[0];
+    }
+};
+
+/**
+ * Shows a dialog component given its IoCSS selector.
+ * @param dialogManager {Component} The `gpii.app.dialogManager` instance.
+ * @param selector {String} The IoCSS selector of the component to be shown.
```
Comment (kaspermarkus) [inline src/main/dialogManager.js:107]: pretty self explanatory, but for good measure, could you add an @param options

File: src/main/factsManager.js [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+var fluid = require("infusion");
+
+var gpii = fluid.registerNamespace("gpii");
+
+require("./utils.js");
+
+/**
+ * Manages and disseminates facts. A fact is simply represented by a name
+ * and value. For each fact there is a fact provider that collects and supplies it.
+ * Example for fact is "keyedInBefore", which defines times since keying in in
```
Comment (kaspermarkus) [inline src/main/factsManager.js]: Example for fact is -> Example of a fact is

File: src/main/rulesEngine.js [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+                "{that}.registeredRulesMap",
+                "{arguments}.0"
+            ]
+        },
+        addRule: {
+            funcName: "gpii.app.rulesEngine.addRule",
+            args: [
+                "{that}",
+                "{that}.registeredRulesMap",
+                "{arguments}.0",
```
Comment (kaspermarkus) [inline src/main/rulesEngine.js]: good practice is to do a comment next to each argument specifying it's name - such as:
"{arguments}.0", // ruleId

It doesn't have to be more than that. These only serves as a quick reminder what it is, for a proper description, one can look in the actual function comments

File: src/main/surveys/surveyConnector.js [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+ * https://github.com/GPII/universal/blob/master/LICENSE.txt
+ */
+"use strict";
+
+var fluid = require("infusion"),
+    gpii = fluid.registerNamespace("gpii");
+
+/**
+ * Send/receive survey data to/from the survey server
+ * occurred
```
Comment (kaspermarkus) [inline src/main/surveys/surveyConnector.js]: dont think this "occurred" was supposed to be there

File: src/main/surveys/surveyConnector.js [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+ * would look like this:
+ *     {
+ *         type: "triggersRequest",
+ *         value: {
+ *             userId: <keyedInUserToken> // the token of the currently keyed in user
+ *             machineId: <machineId> // the installation id of the OS
+ *         }
+ *     }
+ *
+ * When the server receives a "surveyTrigger" request, it will respond with
```
Comment (kaspermarkus) [inline src/main/surveys/surveyConnector.js]: Isn't this supposed to be "receives a "triggersRequest" request"

File: src/main/surveys/surveyConnector.js [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+
+    switch (type) {
+    case "survey":
+        events.onSurveyRequired.fire(value);
+        break;
+    case "surveyTrigger":
+        events.onTriggerDataReceived.fire(value);
+        break;
+    default:
+        console.log("SurveyConnector - Unrecognized message type: ", message);
```
Comment (kaspermarkus) [inline src/main/surveys/surveyConnector.js]: fluid.log

File: src/main/surveys/surveyDialog.js [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+"use strict";
+
+var fluid = require("infusion"),
+    electron = require("electron"),
+    ipcMain = electron.ipcMain,
+    gpii = fluid.registerNamespace("gpii");
+
+require("../dialog.js");
+
+// XXX: Needed to circumvent the certificate error for the "umd.edu" domain.Should
```
Comment (kaspermarkus) [inline src/main/surveys/surveyDialog.js]: Create JIRA for this (and reference)

**Review by kaspermarkus — COMMENTED** [review]
This is generally really nice code, and all my comments are really minor.

Generally, for all instances of {arguments.*} it would be nice to have a comment with the name of the argument to the function (as mentioned in one of my comments) as that saves one from having to look at the actual function to get reminded of what the parameters are.

File: src/main/dialogManager.js (hunk 3 above)
Comment (georgitodorov) [reply src/main/dialogManager.js:33]: I have just created a ticket for this improvement: https://issues.gpii.net/browse/GPII-2817

File: src/main/surveys/surveyDialog.js (hunk 10 above)
Comment (georgitodorov) [reply src/main/surveys/surveyDialog.js]: The ticket for this issue is https://issues.gpii.net/browse/GPII-2818.

File: main.js [hunk 11]
```diff
@@ -35,5 +35,5 @@ require("./src/main/app.js");
 
 kettle.config.loadConfig({
     configName: kettle.config.getConfigName("app.testing"),
-    configPath: kettle.config.getConfigPath(__dirname + "/configs")
+    configPath: kettle.config.getConfigPath(fluid.module.terms()["gpii-app"] + "/configs")
```
Comment (amb26) [inline main.js]: This should just be "%gpii-app/configs"

File: src/main/app.js [hunk 12]
```diff
@@ -313,6 +414,22 @@ fluid.defaults("gpii.app", {
...
 });
 
+/**
+ * Retrieves the id of the machine on which the PSP is running and sets it in the
+ * model. In case the machine id cannot be obtained, this will not prevent the app
+ * from starting.
+ * @param psp {Component} The `gpii.app.psp` component.
+ */
+gpii.app.fetchMachineId = function (that) {
+    machineInfo.machineId().then(function (machineId) {
```
Comment (amb26) [inline src/main/app.js]: Please harmonise with whatever work has been done by @stegru on acquiring machine id in the core

File: package.json [hunk 13]
```diff
@@ -11,22 +11,28 @@
...
+    "start-psp-dev": "electron main.js ./configs app.dev",
+    "dev": "npm-run-all --parallel start-survey-server start-psp-dev",
     "test": "powershell provisioning/Tests.ps1 -originalBuildScriptPath './node_modules/gpii-windows/provisioning'; electron tests.js"
   },
   "dependencies": {
     "electron": "1.4.1",
     "electron-edge": "6.5.4",
     "gpii-windows": "git://github.com/danailbd/windows.git#hst-2017",
     "infusion": "3.0.0-dev.20170830T182157Z.392b2f8",
+    "json-rules-engine": "2.0.2",
```
Comment (amb26) [inline package.json]: Normally when we introduce a significant new library into the GPII, we perform a technology evaluation exercise, which sets out the requirements on the library, surveys a good selection of those available and matches them against the criteria. You can see some examples of evaluations we have done in the past at https://wiki.gpii.net/w/Technology_Evaluation - . In this case, I'm not clear that the chosen technology solves the difficult part of the problem, which is scheduling of events to occur asynchronously when conditions are met in the future, without incurring too much polling. Note that if we are just considering evaluating rules or expressions encoded in a declarative form, the existing model transformation system described at https://docs.fluidproject.org/infusion/development/ModelTransformationAPI.html can do most of what is required, perhaps with the addition of a couple of new transformation rules. Also note that @kaspermarkus  implemented a somewhat similar system in the GPII's existing "ContextManager" at https://github.com/GPII/universal/blob/master/gpii/node_modules/contextManager/src/ContextManager.js - this maintains a "context" (very similar to your collection of "facts") which takes in environmental information, including the current time, etc. and through a polling system causes a change of "context" which is specified by declarative rules such as this one: https://github.com/GPII/universal/blob/master/gpii/node_modules/contextManager/test/js/ContextManagerUtilitiesTests.js#L51

**Comment by danailbd** [comment]
Closing this PR, as it will be redirected to the main repo. The base branch was merged, so there is no need for this PR to be internal anymore.

**Pull request closed without merging**